### ISSUE TITLE
Not all my tests are detected and executed

### ISSUE THREAD

**Issue by AlexandreCantin** [oven-sh/bun#3506]
### What version of Bun is running?

0.6.12

### What platform is your computer?

Darwin 22.5.0 arm64 arm

### Explain the bug :
TL;DR => Some tests are not detected and ignored

When running my test with `bun test`, some of my test are not detected (and not executed)

For example, I have these file at `tests/user/services/slugify.test.ts` :

```
import { slugify } from '$src/user-context/services/slugify';
import { describe, expect, test } from 'bun:test';

describe('slugify service', () => {
	test('should lower-case the text', () => {
		const text = 'Hello World';
		const result = slugify(text);
		expect(result).toBe('hello-world');
	});

	test('should trim the text', () => {
		const text = '  Hello World  ';
		const result = slugify(text);
		expect(result).toBe('hello-world');
	});

	test('should raise for empty text', () => {
		const text = '   ';
		expect(() => slugify(text)).toThrow('Empty slug');
	});
});
```

And when I run `bun test`, no test are detected :

```
alexandre-cantin@macbook-pro-de-alexandre bun % bun test
bun test v0.6.12 (039bbc68)
[0.31ms] ".env"

... other tests

tests/user/use-cases/validate-user.test.ts:

tests/user/services/slugify.test.ts:
==> NO TEST EXECUTED 🫤

 10 pass
 0 fail
 33 expect() calls
Ran 10 tests across 6 files. 10 total [211.00ms]
```
How to be sure it's not a visual bug ?
If I update a test in `slugify.test.ts` in order to fail => `bun test` detects no failing test and say that's all good 

But if I rename the folder `tests/user/services/` to `tests/user/slugify/`, my 3 tests are executed :

```
alexandre-cantin@macbook-pro-de-alexandre bun % bun test
bun test v0.6.12 (039bbc68)
[0.27ms] ".env"

... other tests

tests/user/slugify/slugify.test.ts:
✓ slugify service > should lower-case the text [1.36ms]
✓ slugify service > should trim the text [0.05ms]
✓ slugify service > should raise for empty text [0.04ms]

 13 pass
 0 fail
 36 expect() calls
Ran 13 tests across 6 files. 13 total [205.00ms]
```

On the Github actions CI, it's worse because less test are executed than in local 😅

![Capture d’écran 2023-07-03 à 09 25 49](https://github.com/oven-sh/bun/assets/1846641/a1a2ffac-bccc-4d18-9529-6a572cf9ceb6)

By the way, it's not the only an example in this repo => tests in `tests/user/use-cases/validate-user.test.ts` are not detected or executed either… But they are if I rename `register-user.test.ts` (in the same folder) to `register-user.testss.ts` to make it undetected by `bun:test`…

### To reproduce :
- Clone the repo `https://github.com/AlexandreCantin/bun-template`
- `bun install`
- `bun test`

### What is the expected behavior?

All my test are detected and executed despite the folder structure

### What do you see instead?

```
alexandre-cantin@macbook-pro-de-alexandre bun-template % bun test                                                 
bun test v0.6.12 (039bbc68)

tests/health-check.test.ts:
✓ checking if wiptest works > check server ping [6.41ms]

tests/user/value-objects/encrypted-password.test.ts:
✓ encrypted-password value object > valid case [0.14ms]
✓ encrypted-password value object > error case [0.07ms]

tests/user/value-objects/locale.test.ts:
✓ locale value object > valid case [0.68ms]
✓ locale value object > error case [0.07ms]

tests/user/use-cases/register-user.test.ts:
✓ register-user use case > sanitize fields [8.94ms]
✓ register-user use case > missing fields [0.87ms]
✓ register-user use case > malformed email [0.45ms]
✓ register-user use case > too long email [0.56ms]
✓ register-user use case > send register email [27.80ms]

tests/user/use-cases/validate-user.test.ts:

tests/user/services/slugify.test.ts:

 10 pass
 0 fail
 33 expect() calls
Ran 10 tests across 6 files. 10 total [259.00ms]
```
==> No test in `tests/user/services/slugify.test.ts` are executed… (and in `tests/user/use-cases/validate-user.test.ts`) - despite the fact that the files are correctly detected

### Additional information

The `tests` folder structure seems to impact the execution of the tests (despite the file being detected)

**Comment by Jarred-Sumner** [comment]
Might be a bug specific to having multiple instances of the word “test” or “spec” in the filename. I haven’t checked yet but it reads like we are doing an indexOf instead of a lastIndexOf

**Comment by sv2dev** [comment]
Similar problem here:
`Bun 1.0.14`, running on `Darwin 23.1.0`.
I've my `*.spec.ts` files listed next to the implementation files. Apart from the extension, they don't have a `test` or `spec` within their absolute path name.
Until recently, all tests were detected. In the CI pipeline (within a docker container from image `oven/bun:1.0.14-slim`), more test files are detected than locally.

Output loally: `Ran 7 tests across 11 files.`
Output CI: `Ran 41 tests across 11 files.`

I tried cloning the git repo to another directory, but the same subset of tests is executed.

If I search for `it(` through my project, I find 55 tests within the 11 test files. So it seems that even on the CI system, not all tests are executed.

**Comment by sv2dev** [comment]
Ah, my bad.
I just found a suite marked with`.only()`.
But I still find it confusing that there is a different amount of tests is executed in a different environment.

**Comment by gjovanov** [comment]
Any update on when it is expected to be fixed?

**Comment by jimlloyd** [comment]
I had this problem too. It turns out that even a single .only can cause multiple files to be skipped.

**Comment by Kostas-Xafis** [comment]
Still having the same problem in bun:1.1.29.
I am on a 23.10 Ubuntu distro.

I have 2 files like: 
1) `./tests/api/registrations.test.ts`
2) `./tests/api/teachers.test.ts`

Both get recognized by bun but only the tests from the first file execute. 

Even though I do not make use of any of expect helpers (like `.only()`, `.if()` etc), but only of `.skip()` which I assume that it shouldn't contribute to the issue here.
